src/docs/sphinx/dev_guide/tensor_dual.rst

```diff
@@ -0,0 +1,405 @@
+
+In general, each part of a function's output can depend on each part of
+its inputs. So, in this example the gradient could potentially have up
+to 81 components:
+
+.. math:: \frac{\partial \sigma_{ij}}{\partial \epsilon_{kl}}, \qquad i,j,k,l \in {1,2,3}
+
+If we promote the input argument to a tensor of dual numbers, we can
+compute these derivatives automatically:
+
+.. code-block:: cpp
+
+   tensor< double, 3, 3 > epsilon = {...};
+   tensor< dual< tensor< double, 3, 3 > >, 3, 3 > sigma = stress(make_dual(epsilon));
+
+Now, ``sigma`` contains value and gradient information that can be
+understood in the following way
+
+.. math:: \texttt{sigma[i][j].value} = \sigma_{ij} \qquad \texttt{sigma[i][j].gradient[k][l]} = \frac{\partial \sigma_{ij}}{\partial \epsilon_{kl}}
+
+There are also convenience routines to extract all the values and
+gradient terms into their own tensors of the appropriate shape:
+
+.. code-block:: cpp
+
+   // as before
+   tensor< dual< tensor< double, 3, 3 > >, 3, 3 > sigma = stress(make_dual(epsilon));
+
+   // extract the values
+   tensor< double, 3, 3 > sigma_values = get_value(sigma);
+
+   // extract the gradient
+   tensor< double, 3, 3, 3, 3 > sigma_gradients = get_gradient(sigma);
+
+
+
+Differentiating Functions with Multiple Inputs and Outputs
+-----------------------------------------------------------
+
+Now let's consider a function that has multiple inputs and multiple outputs:
+
+.. code-block:: cpp
+
+   double mu = 1.0;
+   double rho = 2.0;
+   static constexpr auto I = Identity<3>();
+   auto f = [=](auto p, auto v, auto L){ 
+      auto strain_rate = 0.5 * (L + transpose(L));
+      auto stress = - p * I + 2 * mu * strain_rate;
+      auto kinetic_energy_density = 0.5 * p * dot(v, v);
+      return std::tuple{stress, kinetic_energy_density};
+   };
+
+Here, ``f`` calculates the stress, :math:`\sigma`, and local kinetic energy density, :math:`q`, of a fluid in terms of
+the pressure ``p`` (scalar), velocity ``v`` (3-vector), and velocity gradient ``L`` (3x3 matrix).
+So, there are 2 outputs and 3 inputs, resulting in potentially 6 derivatives with different order tensors:
+
+.. math:: 
+
+   \frac{\partial \sigma}{\partial p}, \frac{\partial \sigma}{\partial v}, \frac{\partial \sigma}{\partial L},
+   \frac{\partial q}{\partial p}, \frac{\partial q}{\partial v}, \frac{\partial q}{\partial L}
+
+All of these derivatives can be calculated in a single function invocation by following the same
+pattern as before:
+
+.. code-block:: cpp
+
+   double p = ...;
+   tensor<double,3> v = ...;
+   tensor<double,3,3> L = ...;
+
+   // promote the arguments to dual numbers with make_dual()
+   std::tuple dual_args = make_dual(p, v, L);
+
+   // then call the function with the dual arguments
+   auto outputs = std::apply(f, dual_args);
+
+   // note: std::apply is a way to pass an n-tuple to a function that expects n arguments 
+   // 
+   // i.e. the two following lines have the same effect
+   // f(p, v, L);
```

Comment:
I think it would be better to have these two lines uncommented since they are actual code.